src/amber/cli/helpers/process_runner.cr

```diff
@@ -1,114 +1,233 @@
+require "http"
 require "./helpers"
+require "../config"
+require "./file_watcher"
+require "../../exceptions/exception_page"
 
-module Sentry
-  class ProcessRunner
-    property processes = [] of Process
-    property process_name : String
-    property files = [] of String
-    @logger : Amber::Environment::Logger
-    FILE_TIMESTAMPS = {} of String => String
-
-    def initialize(
-      @process_name : String,
-      @build_command : String,
-      @run_command : String,
-      @build_args : Array(String) = [] of String,
-      @run_args : Array(String) = [] of String,
-      files = [] of String,
-      @logger = Amber::CLI.logger
-    )
-      @files = files
-      @npm_process = false
-      @app_running = false
+module Amber::CLI::Helpers
+  include Environment
+
+  struct ProcessRunner
+    PROCESSES = [] of {Process, String}
+
+    @host : String
+    @port : Int32
+    @file_watcher : FileWatcher
+
+    def initialize
+      @watch_running = false
+      @wait_build = Channel(Bool).new
+      @server_files_changed = false
+      @notify_counter = 0
+      @notify_counter_channel = Channel(Int32).new
+      @notify_channel = Channel(Nil).new
+      @host = Helpers.settings.host
+      @port = Helpers.settings.port
+      @file_watcher = FileWatcher.new
+
+      at_exit do
+        kill_processes
+      end
+
+      Signal::INT.trap do
+        Signal::INT.reset
+        exit
+      end
     end
 
     def run
-      loop do
-        scan_files
-        sleep 1
-      end
-    end
-
-    # Compiles and starts the application
-    def start_app
-      build_result = build_app_process
-      if build_result.is_a? Process::Status
-        if build_result.success?
-          stop_all_processes
-          create_all_processes
-          @app_running = true
-        elsif !@app_running
-          log "Compile time errors detected. Shutting down..."
-          exit 1
+      if watch_config = CLI.config.watch
+        run_watcher(watch_config)
+      else
+        warn "Can't find watch settings, do you want to add default watch settings? (y/n)"
+        if gets.to_s.lowercase == "y"
+          CLI.generate_config
         end
+        exit 1
       end
+    rescue ex : KeyError
+      error "Error in watch configuration. #{ex.message}"
+      exit 1
     end
 
-    private def scan_files
+    private def run_watcher(watch_config)
+      begin
+        server_config = watch_config["server"]
+        spawn watcher("server", server_config["files"], server_config["commands"])
+      rescue ex
+        error "Error in watch configuration. #{ex.message}"
+        exit 1
+      end
+      watch_config.each do |key, value|
```

Comment:
This should be its own method and be able to loop to all config keys including `watch_config["server"]`

Reply:
I applied @robacarp suggestion here, although, the new refactoring won't require this sweat_smile